modules/renter/hostdb/negotiate.go

```diff
@@ -443,36 +443,75 @@ func (hdb *HostDB) newHostUploader(settings modules.HostSettings) (*hostUploader
 	return hu, nil
 }
 
+// A HostPool is a collection of hosts used to upload a file.
+type HostPool interface {
+	// UniqueHosts will return up to 'n' unique hosts that are not in 'old'.
+	UniqueHosts(n int, old []modules.NetAddress) []Uploader
+
+	// Close terminates all connections in the host pool.
+	Close() error
+}
+
+type pool struct {
+	hosts []*hostUploader
+	hdb   *HostDB
+}
+
+func (p *pool) Close() error {
+	for _, h := range p.hosts {
+		h.Close()
+	}
+	return nil
+}
+
 // UniqueHosts will return up to 'n' unique hosts that are not in 'old'. Note
-// that this is a blocking call that performs network I/O.
-func (hdb *HostDB) UniqueHosts(n int, old []Uploader) (hosts []Uploader) {
-	// remove old hosts from tree
-	hdb.mu.Lock()
-	var oldEntries []*hostEntry
-	for _, host := range old {
-		node, exists := hdb.activeHosts[host.Address()]
-		if !exists {
-			continue
+// that this may require negotiating new contracts.
+func (p *pool) UniqueHosts(n int, old []modules.NetAddress) (hosts []Uploader) {
+	if n == 0 {
+		return
+	}
+
+	// first reuse existing connections
+outer:
+	for _, h := range p.hosts {
+		for _, ip := range old {
+			if h.Address() == ip {
+				continue outer
+			}
+		}
+		hosts = append(hosts, h)
+		if len(hosts) >= n {
+			return hosts
 		}
-		node.removeNode()
-		delete(hdb.activeHosts, host.Address()) // probably unnecessary
-		oldEntries = append(oldEntries, node.hostEntry)
-	}
-	// select random hosts from remaining set
-	randHosts := hdb.randomHosts(n)
-	// replace removed hosts
-	for _, entry := range oldEntries {
-		hdb.insertNode(entry)
 	}
-	hdb.mu.Unlock()
 
-	// negotiate contracts with each host
+	// TODO: revise old, unfilled contracts
+
+	// form new contracts from randomly-picked nodes
```

Comment:
I'm increasingly of the opinion that we should not batch multiple files into a single file contract until either:

A. we have a pressing need due to scalability issues
B. we have a good garbage collection strategy.
C. we have a good UX plan for letting users know that the data won't be coming down. An example of this strategy might involve maintaining a git-ian style history of the renter (almost like IPFS) so that deleted or changed files can be reverted to an earlier state. This gives us a good excuse to not garbage collect, but in itself would be a pretty huge undertaking.